aws/resource_aws_s3_bucket_inventory.go

```diff
@@ -0,0 +1,445 @@
+
+func resourceAwsS3BucketInventoryDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	input := &s3.DeleteBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Deleting S3 bucket inventory configuration: %s", input)
+	_, err = conn.DeleteBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			return nil
+		}
+		return fmt.Errorf("Error deleting S3 bucket inventory configuration: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsS3BucketInventoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	d.Set("bucket", bucket)
+	d.Set("name", name)
+
+	input := &s3.GetBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Reading S3 bucket inventory configuration: %s", input)
+	output, err := conn.GetBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			log.Printf("[WARN] %s S3 bucket inventory configuration not found, removing from state.", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.Set("enabled", *output.InventoryConfiguration.IsEnabled)
+	d.Set("included_object_versions", *output.InventoryConfiguration.IncludedObjectVersions)
+	d.Set("optional_fields", flattenStringList(output.InventoryConfiguration.OptionalFields))
+	d.Set("filter", flattenS3InventoryFilter(output.InventoryConfiguration.Filter))
+	d.Set("schedule", flattenS3InventorySchedule(output.InventoryConfiguration.Schedule))
+	d.Set("destination", []map[string]interface{}{
+		{
+			"bucket": flattenS3InventoryS3BucketDestination(output.InventoryConfiguration.Destination.S3BucketDestination),
+		},
+	})
+
+	return nil
+}
+
+func expandS3InventoryFilter(m map[string]interface{}) *s3.InventoryFilter {
+	v, ok := m["prefix"]
+	if !ok {
+		return nil
+	}
+	return &s3.InventoryFilter{
+		Prefix: aws.String(v.(string)),
+	}
+}
+
+func flattenS3InventoryFilter(filter *s3.InventoryFilter) []map[string]interface{} {
+	if filter == nil {
+		return nil
+	}
+
+	result := make([]map[string]interface{}, 0, 1)
+
+	m := make(map[string]interface{}, 0)
+	if filter.Prefix != nil {
+		m["prefix"] = *filter.Prefix
+	}
+
+	result = append(result, m)
+
+	return result
+}
+
+func flattenS3InventorySchedule(schedule *s3.InventorySchedule) []map[string]interface{} {
+	result := make([]map[string]interface{}, 0, 1)
+
+	m := make(map[string]interface{}, 1)
+	m["frequency"] = *schedule.Frequency
```

Comment:
To prevent potential panics, we should prefer to use the SDK provided function for dereferencing values (`aws.StringValue(schedule.Frequency)`) instead of directly dereferencing via `*`.